matrix_sdk_appservice/src/warp.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright 2021 Famedly GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{net::ToSocketAddrs, result::Result as StdResult};
+
+use futures::TryFutureExt;
+use matrix_sdk::Bytes;
+use serde::Serialize;
+use warp::{filters::BoxedFilter, path::FullPath, Filter, Rejection, Reply};
+
+use crate::{Appservice, Error, Result};
+
+pub async fn run_server(
+    appservice: Appservice,
+    host: impl Into<String>,
+    port: impl Into<u16>,
+) -> Result<()> {
+    let routes = warp_filter(appservice);
+
+    let mut addr = format!("{}:{}", host.into(), port.into()).to_socket_addrs()?;
+    if let Some(addr) = addr.next() {
+        warp::serve(routes).run(addr).await;
+        Ok(())
+    } else {
+        Err(Error::HostPortToSocketAddrs)
+    }
+}
+
+pub fn warp_filter(appservice: Appservice) -> BoxedFilter<(impl Reply,)> {
+    warp::any()
+        .and(filters::valid_access_token(appservice.registration().hs_token.clone()))
+        .and(filters::transactions(appservice))
```

Comment:
The `valid_access_token` Filter needs to be nested under the `path!` matchers, as otherwise `valid_access_token` will lead to rejection & recover for unrelated paths